smart-contract-verifier/src/solidity/verifier.rs

```diff
@@ -424,116 +51,245 @@ impl Verifier {
 
-        self.bc_creation_tx_input
-            .verify_bytecode_with_extra_data(&bytecode)?;
+        Self::compare_creation_tx_inputs(&self.remote_bytecode, &local_bytecode)?;
 
-        let constructor_args = self.extract_constructor_args(abi.constructor(), &bytecode)?;
+        let constructor_args = Self::extract_constructor_args(
+            self.remote_bytecode.creation_tx_input(),
+            local_bytecode.creation_tx_input(),
+            abi.constructor(),
+        )?;
 
         Ok((abi.into_owned(), constructor_args))
     }
 
-    /// Checks that solc versions obtained from metadata hash correspond
-    /// for provided deployed bytecode and deployed bytecode obtained
-    /// as a result of local compilation.
-    fn check_metadata_hash_solc_versions(
-        &self,
-        deployed_bytecode: &DeployedBytecode,
-    ) -> Result<(), VerificationError> {
-        let compiled_solc = &deployed_bytecode.metadata_hash().solc;
-        let bc_solc = &self.bc_deployed_bytecode.metadata_hash().solc;
-        if bc_solc != compiled_solc {
-            let compiled_solc = compiled_solc
-                .as_ref()
-                .map(|b| DisplayBytes::from(b.clone()).to_string());
-            let bc_solc = bc_solc
-                .as_ref()
-                .map(|b| DisplayBytes::from(b.clone()).to_string());
-            return Err(VerificationError::CompilerVersionMismatch(Mismatch::new(
-                compiled_solc,
-                bc_solc,
-            )));
+    fn compare_creation_tx_inputs(
+        remote_bytecode: &Bytecode,
+        local_bytecode: &LocalBytecode,
+    ) -> Result<(), VerificationErrorKind> {
+        let remote_creation_tx_input = remote_bytecode.creation_tx_input();
+        let local_creation_tx_input = local_bytecode.creation_tx_input();
+
+        if remote_creation_tx_input.len() < local_creation_tx_input.len() {
+            return Err(VerificationErrorKind::BytecodeLengthMismatch {
+                part: Mismatch::new(
+                    local_creation_tx_input.len(),
+                    remote_creation_tx_input.len(),
+                ),
+                raw: Mismatch::new(
+                    local_creation_tx_input.clone().into(),
+                    remote_creation_tx_input.clone().into(),
+                ),
+            });
+        }
+
+        Self::compare_bytecode_parts(
+            remote_creation_tx_input,
+            local_creation_tx_input,
+            local_bytecode.creation_tx_input_parts(),
+        )?;
+
+        Ok(())
+    }
+
+    /// Performs an actual comparison of locally compiled bytecode
+    /// with remote bytecode provided for verification.
+    ///
+    /// # Panics
+    ///
+    /// The function will panic if `remote_raw.len()` is less than `local_raw.len()`.
+    fn compare_bytecode_parts(
+        remote_raw: &Bytes,
+        local_raw: &Bytes,
+        local_parts: &Vec<BytecodePart>,
+    ) -> Result<(), VerificationErrorKind> {
+        assert!(
```

Comment:
didn't get it, why not `return Err(...)` ?

Reply:
It is not an error. It was added just to make explicit that precondition. It is the responsibility of the caller to ensure thtlat it never fails. I.e. currently it is used only in `compare_creation_tx_inputs` and there is condition that ensures that property. Probably it does worth to add a comment describing that line, thank you!